Comments: add quick actions to the comments list view

rodrigoi opened the pull request:
This PR makes a small modification to `CommentDetailActions` so it can be reused for quick actions on hover.

![oct-09-2017 09-40-02](https://user-images.githubusercontent.com/233601/31338508-ee997a6e-acd5-11e7-9633-4dcd259f5aa2.gif)

Changes include:

- Upgrading `CommentDetailHeader` from stateless component to a `class` so it can hold the `hover` in local state.
- Adding a `is-quick` css rule to `comment-detail__actions`.
- Selectively showing the action buttons as children of a `ButtonGroup` depending on the hover state.
- Change the style and content of the action buttons depending on the hover state.

To test:

- Go to a site's comments.
- Quick actions should show on hover and work as expected.

References:

Fixes #14119.
p3fqKv-5LC-p2

Copons commented:
As mentioned elsewhere: beautiful!

I've got a couple of things I'd like to discuss:

- I know it's an unpopular opinion, but what about handling the visibility of the actions via CSS?
It's unpopular because we end up always rendering them, but we could then avoid displaying them altogether on small screens (on desktop too, as it doesn't look great imho).

- We should `stopPropagation` action clicks to avoid expanding the comment—which is the expected behaviour for the Edit action only, whereas it looks odd both on Like/Approve toggling (the comment stays there and it expands), and on Spam/Trash (the comments expands and immediately after gets unmounted and disappears).

- Like and approve actions should be coloured (yellow-orange-ish and green respectively) if active.

- Super minor nitpicks: the quick actions icons are a tad vertically misaligned, and the quick actions group could use a little more spacing from the collapse arrow.

rodrigoi commented:
> I know it's an unpopular opinion, but what about handling the visibility of the actions via CSS?
It's unpopular because we end up always rendering them, but we could then avoid displaying them altogether on small screens (on desktop too, as it doesn't look great imho).

To be honest, my first instinct to get this out quick was to use local state, but I'll explore the option of using pure CSS.

> We should `stopPropagation` action clicks to avoid expanding the comment...

Yeah, I noticed that and I'm working on fixing it.

> Like and approve actions should be coloured (yellow-orange-ish and green respectively) if active.

There are some style changes already, but none involving color. I'd like to wait for @megsfulton feedback on that.

> Super minor nitpicks: the quick actions icons are a tad vertically misaligned, and the quick actions group could use a little more spacing from the collapse arrow.

It was really annoying, and it took me a while to figure out, but it's fixed on b159bce7ff48b69440066e80979087755720deba.

megsfulton commented:
1) Let's get rid of edit as a quick action for now. I think it's probably an uncommon behavior compared with the other actions and seems okay to have it only accessible through the expanded view. 
2) If we want to expand the quick actions set beyond the original proposal of approve, spam and trash I'd argue that we probably want reply in there as well. And when reply is clicked it would expand the box with the reply field focused.
3) Agree w/ @Copons regarding color. Let's add the green and yellow for like and approve. 
4) I just realized we're using two different icons for approve depending on the comment state. I think we should use the filled circle with a check everywhere. 
5) You might already know about it but there's something wonky going on with the alignment in the expanded view. Screenshot from chrome. 
<img width="435" alt="screen shot 2017-10-09 at 11 47 24 am" src="https://user-images.githubusercontent.com/161697/31353510-b24d0e38-ace7-11e7-8d60-4e7d7a0e4d54.png">

Also looking forward to seeing the transition / expand updates that are in the works, hard to get a feel for it right now.

megsfulton commented:
Also checked this out on mobile and looks like the hover state is interfering with the tapped state.
![img_9965](https://user-images.githubusercontent.com/161697/31354465-5a30fa9e-aceb-11e7-9c03-a52b61ed2bb1.jpg)

roundhill commented:
Looking good! Maybe we could add a hover tooltip to the buttons since we aren't using labels for these?

Copons commented:
> 1) Let's get rid of edit as a quick action for now. I think it's probably an uncommon behavior compared with the other actions and seems okay to have it only accessible through the expanded view. 
> 2) If we want to expand the quick actions set beyond the original proposal of approve, spam and trash I'd argue that we probably want reply in there as well. And when reply is clicked it would expand the box with the reply field focused.

Second these!
Although, maybe 2) could use its own PR, because it adds a new feature to the actions.

> 4) I just realized we're using two different icons for approve depending on the comment state. I think we should use the filled circle with a check everywhere. 

We should consider that Approved and Like can be active and inactive in the same lists (as opposed to Trash and Spam that are only active in their respective lists).
Of course the color helps there, but I wonder (from a completely uninformed point of view) if it's enough for color blind people.

> Also checked this out on mobile and looks like the hover state is interfering with the tapped state.
> ![img_9965](https://user-images.githubusercontent.com/161697/31354465-5a30fa9e-aceb-11e7-9c03-a52b61ed2bb1.jpg)

I wish to reiterate that imho quick actions don't work on small screens at all.
Not only because we might assume that in most cases `small screens = touch`, but also because we risk covering too much of the actual comment card, obstructing its clickable area.

megsfulton commented:
>We should consider that Approved and Like can be active and inactive in the same lists (as opposed to Trash and Spam that are only active in their respective lists).
Of course the color helps there, but I wonder (from a completely uninformed point of view) if it's enough for color blind people.

That's fair. Just wish the regular check and filled circle check were closer to one another, like how the star is filled/unfilled. Let's leave as is but introduce the colors. 

>I wish to reiterate that imho quick actions don't work on small screens at all.
Not only because we might assume that in most cases small screens = touch, but also because we risk covering too much of the actual comment card, obstructing its clickable area (and, of course, covering most of the text).

Agree. Are we able to detect the device and disable hover if  the user is on a tablet / mobile device? 
I'm working to get the actions to always be in the card but it's going to require changing the layout of the collapsed view so this would just be temporary.

kwight commented:
I still like the idea of quick actions for mobile, but as distinct buttons at the bottom of each comment card (which, of course, takes up even more room on smaller screens 🙃).

rodrigoi commented:
@Automattic/lannister for everyone's consideration, I've fixed various quirks, made small improvements here and there, fixed the mobile problems and added a reply button.

@Copons to make the reply button work, I had to move the control of the reply box to `CommentDetail`. Would you mind double checking I didn't messed up the height calculations and everything works as expected?

Copons commented:
The `hasFocus` change seems to be working as expected. 👍 

I think you might have missed [this comment](https://github.com/Automattic/wp-calypso/pull/18657#issuecomment-335543357) regarding the approved icon, though.

Aside from that, this all LGTM! 🎉

megsfulton commented:
Could we switch up the order to be: reply, like, approve, spam, trash? 

I think it's important to keep the approve/spam/trash actions together.

kwight commented:
@Copons The strongest value I see in them is for users that manage comments every day (of which I am not one); and for users that _don't_ manage comments every day, the only downside seems to be the clutter you mention (which we've mitigated on smaller screens, at least). Simple and effective is hard 🙃

kwight commented:
> Just wish the regular check and filled circle check were closer to one another, like how the star is filled/unfilled. Let's leave as is but introduce the colors.

Ha, bummer, I was just going to leave a comment that we're using two different checkmark icons 😜  I prefer the same icon (@copons if the argument is accessibility, wouldn't that apply to spam and trash too?), but we don't need to hold up this PR to continue that discussion.

Copons commented:
> if the argument is accessibility, wouldn't that apply to spam and trash too?

@kwight Spam and Trash can't have both active and inactive states in the same list.
Say you are in the Approved list, and hit Spam on a comment: that comment immediately disappears from the Approved list. The same goes when you're in the Spam list and hit Approved.
So in any given list, you can only see all comments being either "spam active", or "spam inactive".

But, when you're in the All, Pending, and Approved lists, due to the "persistent" logic, you can see both approved and pending comments, and consequently the Approve icon can be both gray and green.
In this case it's better imho to have as much distinction possible on the Approve/Unapprove icons.

vindl commented:
Sorry for being late to the party, but the current layout of this with it appearing directly over text (on longer comments) makes it look buggy to me. What do you all think about maybe fading the comment content in the background or adding some dark overlay on hover? This would also allow us to make these controls a lot larger (up to a full row), although I'm not sure how it would work out on touch devices.

rodrigoi commented:
@vindl check out the next iteration here: p3fqKv-5Sx-p2 we'll make them always visible on pending comments and they'll have their own space.

gwwar commented on client/blocks/comment-detail/comment-detail-actions.jsx:
```diff
@@ -13,21 +13,33 @@ import { includes } from 'lodash';
+		</div>
+	) : (
+		<div className="comment-detail__actions">{ children }</div>
+	);
```
Minor point, but can we clean this up later to stick with one component per file? Another option would be to have CommentDetailActions extend Component, to give us more flexibility with helper methods.
